examples/linear_model/plot_poisson_regression_non_normal_loss.py

```diff
@@ -0,0 +1,455 @@
+# study for the sake of simplicitly.
+#
+# We also subsample the data for the sake of computational cost and running
+# time. Using the full dataset would lead to similar conclusions.
+
+df = load_mtpl2(n_samples=300000)
+
+# Correct for unreasonable observations (that might be data error)
+df["Exposure"] = df["Exposure"].clip(upper=1)
+
+##############################################################################
+# The remaining columns can be used to predict the frequency of claim events.
+# Those columns are very heterogeneous with a mix of categorical and numeric
+# variables with different scales, possibly very unevenly distributed.
+#
+# In order to fit linear models with those predictors it is therefore
+# necessary to perform standard feature transformations as follows:
+
+log_scale_transformer = make_pipeline(
+    FunctionTransformer(np.log, validate=False),
+    StandardScaler()
+)
+
+linear_model_preprocessor = ColumnTransformer(
+    [
+        ("passthrough_numeric", "passthrough",
+            ["BonusMalus"]),
+        ("binned_numeric", KBinsDiscretizer(n_bins=10),
+            ["VehAge", "DrivAge"]),
+        ("log_scaled_numeric", log_scale_transformer,
+            ["Density"]),
+        ("onehot_categorical", OneHotEncoder(),
+            ["VehBrand", "VehPower", "VehGas", "Region", "Area"]),
+    ],
+    remainder="drop",
+)
+
+##############################################################################
+# The number of claims (``ClaimNb``) is a positive integer that can be modeled
+# as a Poisson distribution. It is then assumed to be the number of discrete
+# events occurring with a constant rate in a given time interval
+# (``Exposure``, in units of years). Here we model the frequency
+# ``y = ClaimNb / Exposure``, which is still a (scaled) Poisson distribution,
+# and use ``Exposure`` as ``sample_weight``.
+
+df["Frequency"] = df["ClaimNb"] / df["Exposure"]
+
+print(
+   pd.cut(df["Frequency"], [-1e-6, 1e-6, 1, 2, 3, 4, 5]).value_counts()
+)
+
+print("Average Frequency = {}"
+      .format(np.average(df["Frequency"], weights=df["Exposure"])))
+
+print("Percentage of zero claims = {0:%}"
+      .format(df.loc[df["ClaimNb"] == 0, "Exposure"].sum() /
+              df["Exposure"].sum()))
+
+##############################################################################
+# It is worth noting that 92 % of policyholders have zero claims, and if we
+# were to convert this problem into a binary classification task, it would be
+# significantly imbalanced.
+#
+# To evaluate the pertinence of the used metrics, we will consider as a
+# baseline a "dummy" estimator that constantly predicts the mean frequency of
+# the training sample.
+
+df_train, df_test = train_test_split(df, random_state=0)
+
+dummy = make_pipeline(
+    linear_model_preprocessor,
+    DummyRegressor(strategy='mean')
+)
+dummy.fit(df_train, df_train["Frequency"],
+          dummyregressor__sample_weight=df_train["Exposure"])
```

Comment:
For curiosity, Is there a use case for not setting `sample_weight` to the exposure?

Reply:
Yes, but not in this dataset. If all policies had `Exposure=1`, it would not be necessary. Same if you would model the circumference of tree stumps (maybe Gamma) and the data contained the age of tree, the area the rainfall etc. per tree. In the end it depends on the data and the problem at hand.

Reply:
For this specific dataset, one must set the `sample_weight` otherwise they would have incorrect results. Is this correct?

The literature uses "offset" and "exposure" to mean this, and we are adding "sample_weight" to mean this as well. My concern is how users may get incorrect results because it may not be obvious that "sample_weight" means "exposure". Yes we have documented this in the user guide and have an example about it. If setting the exposure is really important for getting correct results, I would even consider highlighting it in the docstring.

BTW this is not a blocker. I am thinking of how one can use this incorrectly.

Reply:
Yes, you are correct. Without exposure, one would get "incorrect" results. Also yes, in insurance literature and software, the sample weights are often dealt with as offset, which works for Poisson with a log-link (=equivalent to sample weights), but otherwise this is not equivalent (=is not correct). Therefore, I prefer and encourage the usage of sample weights to deal with exposure.

Offsets would we a nice to have feature as one can feed the GLM with apriori known effects. Maybe for the future...

Reply:
> Yes, you are correct. Without exposure, one would get "incorrect" results.

Are you sure about it? In my understanding using exposure is a way to weight data points that are more reliable (i.e. have longer observation duration) higher. Without sample weight, and with enough data (i.e. mostly under-fitting with a linear model) I would still expect the model to converge roughly to a similar solution (or at least it wouldn't be completely wrong). Maybe I'm missing something though.  

The same idea could apply to any kind of dataset where the target is some cumulative variable, normalized by time (i.e. number of visits/month). One could argue that sample_weight should be user "age" (i.e. exposure) in the application, since the target is more accurate than. However one can also argue that the behavior of a user over 5-10 years typically changes a lot, and computing averages over such long duration is also not very meaningful. Moreover the user metadata (e.g. actual age, location etc) are taken only at 1 point in time, meaning that large time history can actually lead to less accurate point. In which case, such data points shouldn't be weighted too much. Just saying that in that use case it would be difficult for me to define "correct" or "incorrect" sample_weight.

> Yes we have documented this in the user guide and have an example about it. If setting the exposure is really important for getting correct results, I would even consider highlighting it in the docstring.

Indeed, but that's very specific to this dataset (or to insurance models). Defining or mentioning "exposure" in the user guide of docstrings is really out of scope in my option. We could have a general discussion of what to use as sample_weight in models in general, but I think it's a wider discussion than GLMs.

Reply:
> Without sample weight, and with enough data ... I would still expect the model to converge roughly to a similar solution.

You are right. I should always said "correct" and "incorrect". There are however the following 2 points:
- If a GLM uses the so called canonical link, e.g. Poisson with log-link, and has the intercept term, then it fulfills a **balance property** on the training set: `sum_i sample_weight_i (y_i - y_pred_i)= 0`.
- If the sample weights vary a lot, e.g. one or two order of magnitudes, the convergence to similar results is very slow.